fix(error): error handling for Git repositories with no commits (Fixes #3544)

SkinnnyJay opened the pull request:
## What?
This PR fixes issue #3544 by adding proper error handling for Git repositories that have been initialized but contain no commits yet.

## Why?
Previously, when users tried to use Codex in a Git repository with no commits, they would encounter vague error messages later in the process. This created a poor user experience and made it difficult to understand what was wrong.

## How?
The fix introduces a new helper function `git_repo_has_commits()` in `codex-core` that checks if a Git repository has at least one commit. This function is then used in two key places:

1. **CLI (`codex-exec`)**: Added an early check that provides a clear, actionable error message when the repository has no commits
2. **MCP Server**: Added the same validation during conversation creation to prevent task failures

### Changes Made:
- **`codex-rs/core/src/git_info.rs`**: 
  - Added `git_repo_has_commits()` function that uses `git rev-parse --verify HEAD` to detect repositories without commits
  - Added comprehensive tests for both empty repositories and repositories with commits
  
- **`codex-rs/exec/src/lib.rs`**: 
  - Added validation check that shows clear error message: "This Git repository has no commits. Initialize it first: git add -A && git commit -m 'Initial commit'"
  - Respects the existing `--skip-git-repo-check` flag to allow bypassing this check
  
- **`codex-rs/mcp-server/src/codex_message_processor.rs`**: 
  - Added the same validation during conversation creation
  - Returns proper JSON-RPC error with actionable message
  
- **`codex-rs/tui/src/app.rs`**: 
  - Added test configuration to handle macOS seatbelt restrictions

### Error Messages:
The new error messages are clear and actionable:
- CLI: `"This Git repository has no commits. Initialize it first:\n  git add -A && git commit -m 'Initial commit'\nOr pass --skip-git-repo-check to bypass."`
- MCP: `"This Git repository has no commits. Codex requires at least one commit. Run: git add -A && git commit -m 'Initial commit'"`

### Testing:
- Added unit tests for the new `git_repo_has_commits()` function
- Tests cover both empty repositories and repositories with commits
- All existing tests continue to pass

This change improves the user experience by providing immediate, clear feedback when attempting to use Codex in an uninitialized Git repository.

bolinfest commented on codex-rs/tui/src/app.rs:
```diff
@@ -532,6 +532,10 @@ mod tests {
         }
     }
 
+    #[cfg_attr(
```
How is this related to this PR?

bolinfest commented on codex-rs/exec/src/lib.rs:
```diff
@@ -188,6 +188,17 @@ pub async fn run_main(cli: Cli, codex_linux_sandbox_exe: Option<PathBuf>) -> any
 
+    // If inside a git repo but with no commits, provide a clearer message.
+    if !skip_git_repo_check
+        && get_git_repo_root(&config.cwd.to_path_buf()).is_some()
```
Can you unify this with the code on 186 so it's clear we only run `get_git_repo_root()` once?

An option is to change `get_git_repo_root()` to return an `enum` with three states: Yes, YesButNoCommits, No, though I don't know if that's really necessary if we consolidate this code with the above and change `git_repo_has_commits()` to invoke `git` once.

bolinfest commented on codex-rs/core/src/git_info.rs:
```diff
@@ -108,6 +108,26 @@ pub async fn collect_git_info(cwd: &Path) -> Option<GitInfo> {
+/// commit. Returns false for non-git directories and for repositories with no
+/// commits yet.
+pub async fn git_repo_has_commits(cwd: &Path) -> bool {
+    // Must be in a git repo first
+    let is_git_repo = match run_git_command_with_timeout(&["rev-parse", "--git-dir"], cwd).await {
+        Some(out) => out.status.success(),
+        None => false,
+    };
+    if !is_git_repo {
+        return false;
+    }
```
Do we need this code? Can we just run the second part?

SkinnnyJay committed:
feat(core): add git_repo_has_commits function with comprehensive tests

- Add git_repo_has_commits() function to detect repositories without commits
- Uses git rev-parse --verify HEAD to check for commits after confirming git repo
- Returns false for non-git directories and empty repositories
- Returns true for repositories with at least one commit
- Add comprehensive tests covering all scenarios:
  - Empty git repositories (no commits)
  - Repositories with commits
  - Non-git directories

Addresses: Issue #3544 - Need clear error for empty git repositories

SkinnnyJay committed:
feat(exec): add git commit validation with clear error message

- Add import for git_repo_has_commits function
- Add validation check after existing git repo check
- Provide clear, actionable error message for empty repositories
- Respects existing --skip-git-repo-check flag to bypass validation
- Error message guides users to initialize repository or bypass check

Addresses: Issue #3544 - CLI should show clear error for empty git repos

SkinnnyJay committed:
feat(mcp-server): add git commit validation for conversation creation

- Add imports for get_git_repo_root and git_repo_has_commits functions
- Add validation check during new conversation creation
- Provide clear JSON-RPC error message for empty git repositories
- Returns INVALID_REQUEST error code with actionable message
- Prevents vague task creation failures by catching issue early

Addresses: Issue #3544 - MCP server should show clear error for empty git repos

SkinnnyJay committed:
Merge branch 'main' into bug-3544

SkinnnyJay committed:
Merge branch 'main' into bug-3544

SkinnnyJay committed:
Merge branch 'main' into bug-3544
